### ISSUE TITLE
Right-click menu in queue doesn't stay open

### ISSUE THREAD

**Issue by yang-musa** [nukeop/nuclear#1710]
**Platform:**
Fedora 40 Workstation (Gnome, Wayland), fully updated as of 9/25/24

**Nuclear version:**
0.6.38 installed from Flathub

**Description of the issue:**
Steps to reproduce:

1. Search or navigate to an album or playlist
2. Select Play to populate queue on right
3. If the first search result for a song is incorrect, right-click on song to bring up menu to try alternative search results
4. right-click menu flickers up but doesn't stay open. It's impossible to select new songs from the dropdown menu
5. This bug is slightly inconsistent. Very occasionally it DOES work as intended. But it's rare, almost always it flickers and I can't use it.

Please let me know if there's additional info I can provide to help resolve.

Thanks as always for a great app!

**Comment by nukeop** [comment]
Hey, thanks for reporting. Is it happening when there's other stuff in the queue? I think this is happening when other songs have their streams loaded, too many components are re-rendered then.

**Comment by yang-musa** [comment]
I usually load a full album, so yes there would be 7-12 (?) other songs in the queue.  

As a test I tried emptying the queue and then adding one song, and then trying to right-click. You're right - it worked as expected. So, there's something when there are multiple songs in the queue.

**Comment by nukeop** [comment]
Great, so this issue has been also reported by one more person, and I noticed it myself, pretty sure i know how to fix that so it shouldn't be long.

**Comment by yang-musa** [comment]
I tested adding songs one at a time to see where the limit is. It worked fine until one of the songs was unable to load (the circular loading icon just keeps spinning). Then the right-click doesn't work.

**Comment by nukeop** [comment]
In the meantime you can just wait until all the songs in the queue finish loading, then the menu should be fine.

**Comment by yang-musa** [comment]
Sometimes the queue never finishes loading... See example. In that case there's no way to correct other songs. 
![Screenshot from 2024-09-26 11-05-10](https://github.com/user-attachments/assets/44d1407c-c548-427c-9d5e-2cfb5dfe5de1)

Anyway, sounds like a fix is on the way, I can be patient :-) 

Thanks!

**Comment by waldo121** [comment]
> Hey, thanks for reporting. Is it happening when there's other stuff in the queue? I think this is happening when other songs have their streams loaded, too many components are re-rendered then.

Yes, feels like the whole queue re-renders whenever 1 of the element changes.

**Comment by waldo121** [comment]
@nukeop if you're not working on this already, mind sharing how you thought of fixing this?

**Comment by nukeop** [comment]
This is caused by the entire queue component re-rendering whenever anything in the `queue` subtree of Redux changes. The solution is to prevent these re-renders, either by picking parts of the Redux tree more selectively (e.g. individual songs in each queue item), or by memoizing something.

**Comment by nukeop** [comment]
Also to prevent errored out tracks from blocking the rest of the queue, it would be good to have a "locked with error" state where Nuclear won't make any further attempts at fetching the stream (this can happen when Youtube blocks playback on third party websites).

**Comment by waldo121** [comment]
> This is caused by the entire queue component re-rendering whenever anything in the `queue` subtree of Redux changes. The solution is to prevent these re-renders, either by picking parts of the Redux tree more selectively (e.g. individual songs in each queue item), or by memoizing something.

I have another suggestion since I am not sure that fix would be enough to solve the problem(not sure I fully get yet how redux works). We could move the stream selection ui out of the popup and into the right panel (next to stream verification or at the top of the queue for example). Stream selection might need rethinking then.

**Comment by akashskypatel** [comment]
I noticed that Bandcamp plugin doesnt go into this infinite loading loop. Instead of getStreamForId returning a promise it returns StreamData so if the StreamData is invalid the app just purges the track from the queue. Can other plugins be modified the same way to avoid the infinite loop issue?

I am trying to test this in a local dev build but having issues with getting the source code running locally.

**Comment by nukeop** [comment]
About to be fixed by #1764

**Comment by nukeop** [comment]
- The menu stays open now as the rest of the queue loads
- Nuclear will now correctly try streams one by one, if loading fails. Previously, if the first stream would not load, Nuclear removed the track from the queue. This still happens, but only if we go through all the streams (usually 10-20) without finding a single working one.
- If a queue item is in an error state, we no longer keep reloading it endlessly. You can click it to do so manually.
